// eva-utilities/utils/filter-builder-mysql.js
const validations = {
    text: {
        contains: (field, value, params) => {
            params.push(`%${value}%`);
            return `${field} LIKE ?`;
        },
        notContains: (field, value, params) => {
            params.push(`%${value}%`);
            return `${field} NOT LIKE ?`;
        },
        startsWith: (field, value, params) => {
            params.push(`${value}%`);
            return `${field} LIKE ?`;
        },
        endsWith: (field, value, params) => {
            params.push(`%${value}`);
            return `${field} LIKE ?`;
        },
        equals: (field, value, params) => {
            params.push(value);
            return `${field} = ?`;
        },
        notEqual: (field, value, params) => {
            params.push(value);
            return `${field} != ?`;
        },
        blank: (field) => `${field} IS NULL`,
        notBlank: (field) => `${field} IS NOT NULL`,
    },
    number: {
        equals: (field, value, params) => {
            params.push(value);
            return `${field} = ?`;
        },
        notEqual: (field, value, params) => {
            params.push(value);
            return `${field} != ?`;
        },
        greaterThan: (field, value, params) => {
            params.push(value);
            return `${field} > ?`;
        },
        greaterThanOrEqual: (field, value, params) => {
            params.push(value);
            return `${field} >= ?`;
        },
        lessThan: (field, value, params) => {
            params.push(value);
            return `${field} < ?`;
        },
        lessThanOrEqual: (field, value, params) => {
            params.push(value);
            return `${field} <= ?`;
        },
        between: (field, filter, filterTo, params) => {
            params.push(filter, filterTo);
            return `${field} BETWEEN ? AND ?`;
        },
        blank: (field) => `${field} IS NULL`,
        notBlank: (field) => `${field} IS NOT NULL`,
    },
    date: {
        equals: (field, value, params) => {
            params.push(value);
            return `${field} = ?`;
        },
        notEqual: (field, value, params) => {
            params.push(value);
            return `${field} != ?`;
        },
        after: (field, value, params) => {
            params.push(value);
            return `${field} > ?`;
        },
        lessThan: (field, value, params) => {
            params.push(value);
            return `${field} < ?`;
        },
        inRange: (field, filter, filterTo, params) => {
            params.push(filter, filterTo);
            return `${field} BETWEEN ? AND ?`;
        },
        blank: (field) => `${field} IS NULL`,
        notBlank: (field) => `${field} IS NOT NULL`,
    }
};

const buildFilters = (filters, params = []) => {
    const queryParts = [];
    for (const [field, filter] of Object.entries(filters)) {
        if (!filter.type) {
            let subQueryParts = [];
            for (const condition of filter.conditions) {
                let query;
                if (condition.type === 'between' || condition.type === 'inRange') {
                    query = validations[condition.filterType][condition.type](field, condition.filter, condition.filterTo, params);
                } else {
                    query = validations[condition.filterType][condition.type](field, condition.filter, params);
                }
                subQueryParts.push(`${query}`);
            }
            queryParts.push(`(${subQueryParts.join(` ${filter.operator} `)})`);
        } else {
            let query;
            if (filter.type === 'between' || filter.type === 'inRange') {
                query = validations[filter.filterType][filter.type](field, filter.filter, filter.filterTo, params);
            } else {
                query = validations[filter.filterType][filter.type](field, filter.filter, params);
            }
            queryParts.push(`(${query})`);
        }
    }
    return queryParts.join(" AND ");
};

const generateFilterQuery = (filters) => {
    if (!filters || typeof filters !== "object") {
        return {
            query: "",
            params: [],
        };
    }

    const params = [];
    const whereClause = buildFilters(filters, params);

    return {
        query: whereClause ? `WHERE ${whereClause}` : "",
        params,
    };
};

module.exports = {
    generateFilterQuery,
};
